### ISSUE TITLE
Observing "Exception: startOffset must refer to an offset beyond the Segment's StorageLength offset" in the Segment store while running IO

### ISSUE THREAD

**Issue by vedanthh** [pravega/pravega#3695]
**Observing "Exception: startOffset must refer to an offset beyond the Segment's StorageLength offset" in the Segment store while running IO**

**Environment details: PKS / K8 with medium cluster:**
```
3 master: xlarge: 4 CPU, 16 GB Ram, 32 GB Disk
5 worker: 2xlarge: 8 CPU, 32 GB Ram, 64 GB Disk
Tier-1 storage is from VSAN datastore
Tier-2 storage curved on NFS Client provisioner using Isilon as backend
```
```
Pravega version: 0.5.0-2210.81867f3
Zookeeper Operator : pravega/zookeeper-operator:0.2.1
Pravega Operator: adrianmo/pravega-operator:0.3.2
```
**Segmentstore log error:** 
[Segmentstore_issue_3695.log](https://github.com/pravega/pravega/files/3115538/Segmentstore_issue_3695.log)

```
2019-04-24 21:11:29,559 56072805 [core-23] ERROR i.p.s.server.writer.StorageWriter - StorageWriter[27]: Iteration[156659].Error.
java.lang.IllegalArgumentException: startOffset must refer to an offset beyond the Segment's StorageLength offset.
        at com.google.common.base.Preconditions.checkArgument(Preconditions.java:141)
        at io.pravega.segmentstore.server.reading.StreamSegmentReadIndex.readDirect(StreamSegmentReadIndex.java:626)
        at io.pravega.segmentstore.server.reading.ContainerReadIndex.readDirect(ContainerReadIndex.java:184)
        at io.pravega.segmentstore.server.writer.StorageWriterFactory$StorageWriterDataSource.getAppendData(StorageWriterFactory.java:146)
        at io.pravega.segmentstore.server.writer.SegmentAggregator.getFlushArgs(SegmentAggregator.java:820)
        at io.pravega.segmentstore.server.writer.SegmentAggregator.flushPendingAppends(SegmentAggregator.java:754)
        at io.pravega.segmentstore.server.writer.SegmentAggregator.lambda$flushFully$11(SegmentAggregator.java:665)
        at io.pravega.common.concurrent.Futures$Loop.call(Futures.java:712)
        at io.pravega.common.concurrent.Futures$Loop.call(Futures.java:681)
        at io.pravega.common.concurrent.Futures.runOrFail(Futures.java:572)
        at io.pravega.common.concurrent.Futures$Loop.run(Futures.java:725)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
```

**Comment by RaulGracia** [comment]
@vedanthh this issue could be solved already with PR #3668, but we need to verify this. Could you please check if the issue still persists with `master`?

**Comment by sumit-bm** [comment]
@RaulGracia As we are testing now 0.5.0 would you like to cherry pick this fix in 0.5.0 recent build? Master is in 0.6.0 and i would like not to take it now.

**Comment by RaulGracia** [comment]
@sumit-bm you mean to cherry pick that PR for the `0.5` branch? I think we should port this fix, along with others that are being contributed since the `0.5` tag. I think that we need to keep track of the bugfixes being contributed recently to consider them as candidates for cherry-pick.

**Comment by andreipaduroiu** [comment]
We should port this into 0.5.

**Comment by sumit-bm** [comment]
Thanks @RaulGracia @andreipaduroiu

**Comment by andreipaduroiu** [comment]
fixed with #3668

**Comment by vedanthh** [comment]
Observing this issue with `0.5.0-2236.5228e2d` as well hence reopening 

```
2019-05-03 04:00:53,730 73814816 [core-3] ERROR i.p.s.server.writer.StorageWriter - StorageWriter[28]: Iteration[542431].Error.
java.lang.IllegalArgumentException: startOffset must refer to an offset beyond the Segment's StorageLength offset.
        at com.google.common.base.Preconditions.checkArgument(Preconditions.java:141)
        at io.pravega.segmentstore.server.reading.StreamSegmentReadIndex.readDirect(StreamSegmentReadIndex.java:626)
        at io.pravega.segmentstore.server.reading.ContainerReadIndex.readDirect(ContainerReadIndex.java:184)
        at io.pravega.segmentstore.server.writer.StorageWriterFactory$StorageWriterDataSource.getAppendData(StorageWriterFactory.java:146)
        at io.pravega.segmentstore.server.writer.SegmentAggregator.getFlushArgs(SegmentAggregator.java:820)
        at io.pravega.segmentstore.server.writer.SegmentAggregator.flushPendingAppends(SegmentAggregator.java:754)
        at io.pravega.segmentstore.server.writer.SegmentAggregator.lambda$flushFully$11(SegmentAggregator.java:665)
        at io.pravega.common.concurrent.Futures$Loop.call(Futures.java:712)
        at io.pravega.common.concurrent.Futures$Loop.call(Futures.java:681)
        at io.pravega.common.concurrent.Futures.runOrFail(Futures.java:572)
        at io.pravega.common.concurrent.Futures$Loop.run(Futures.java:725)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
```

**Comment by sumit-bm** [comment]
@vedanthh observed the same exception during IO workload test on build 2236 hence reopening

**Comment by andreipaduroiu** [comment]
This PR hasn't been merged into 0.5 yet. See #3765
